0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,277 @@
+Header-A: some_stuff
+Header-B: some_other_stuff
+Header-C: some_more_stuff
+Date: 2017-03-13T16:57:56.901Z
+Content-Type: application/json
+
+{"foo":"bar"}
+```
+
+Only the payment's sender and receiver can decrypt and read the private headers and their data. The private headers contains an arbitrary message body preceded by the following headers:
+
+| Header | Value |
+|--------|-------|
+| `Expires-At` | _(Optional)_ An [ISO8601 UTC timestamp](https://en.wikipedia.org/wiki/ISO_8601), after which this PSK payment has expired. |
+| `Content-Type` | _(Optional)_ The [media type](http://www.iana.org/assignments/media-types/media-types.xhtml) of the data in the message body |
+| ... | _(Optional)_ Additional arbitrary headers. Where appropriate, it is RECOMMENDED to match the headers defined for HTTP messages ([RFC7230](https://tools.ietf.org/html/rfc7230#section-3.2)) |
+
+## Pseudocode
+
+The following pseudocode outlines the generation procedure for the necessary values involved in PSK. It follows only the case where encryption is enabled, for simplicity.
+
+### Receiver: Setup of the Protocol
+
+This pattern of generating the shared secret based on a random token is not part of the PSK protocol, but allows a receiver to listen for incoming PSK payments with many different shared secrets. If the derivation of the shared secret in this first section is changed, then the [final section](#receiver-receipt-of-each-payment) must be changed to use the same shared secret.
+
+```js
+/**
+  * @param account (this is the address the receiver will listen on, as
+  * returned by their ledger plugin)
+  */ 
+
+// The token is attached to the receiver's ILP address, allowing them to
+// regenerate the shared secret every time they receive a payment, instead of
+// keeping a store of all shared secrets they've given out.
+
+receiver_secret = random_bytes(32)
+token = random_bytes(16)
+
+// The receiver id differentiates this receiver from other receivers on the
+// same account with different receiver_secrets. When they receive a payment,
+// the receiver can make sure the receiver id in the ILP address is their own.
+
+receiver_id = hmac_sha_256(receiver_secret, "ilp_psk_receiver_id").slice(0, 8)
+receiver_address = account + "." + base64url(receiver_id) + base64url(token)
+
+// This pattern of taking the HMAC of a secret and a hard-coded string will be
+// repeated for the generation of many variables. It helps secure the secrets
+// against any interactions between the algorithms used. This method of
+// generating the shared secret is just a recommendation, but the receiver MUST
+// be able to access or regenerate the shared secret in order to fulfill
+// incoming payments.
+
+shared_secret_generator = hmac_sha_256(receiver_secret, "ilp_psk_generation")
+shared_secret = hmac_sha_256(shared_secret_generator, token)
+
+// Now the shared_secret and the receiver_address are transmitted to the
+// sender.
+```
+
+### Sender: Generation of Each Payment
```

Comment:
Ok fine, we can keep it the way it is. It is the comments that make it look long, but it's probably better to have more detail than not.